edc-extensions/bpn-validation/bpn-validation-core/src/main/java/org/eclipse/tractusx/edc/validation/businesspartner/functions/BusinessPartnerGroupFunction.java

```diff
@@ -120,20 +120,23 @@ public boolean evaluate(Operator operator, Object rightValue, Permission rule, P
             return false;
         }
 
-
         var bpn = participantAgent.getIdentity();
         var groups = store.resolveForBpn(bpn);
 
+        var isPositiveOperator = POSITIVE_OPERATORS.contains(operator);
+        var errorMessage = "";
+
         // BPN not found in database
         if (groups.failed()) {
-            policyContext.reportProblem(groups.getFailureDetail());
-            return false;
-        }
-        var assignedGroups = groups.getContent();
+            errorMessage = groups.getFailureDetail();
 
         // BPN was found, but it does not have groups assigned.
-        if (assignedGroups.isEmpty()) {
-            policyContext.reportProblem("No groups were assigned to BPN " + bpn);
+        } else if (groups.getContent().isEmpty()) {
```

Comment:
this code can be simplified, since an empty result is only an error if the operator is "positive", so your if-else cascades should reflect that in the form of a nested if-else: 
```java
if(result.isFailed()){
  policyContext.reportProblem(result.getFailureDetail());
  return false;
}
var groups = result.getGroups();
if (isPositiveOperator && groups.isEmpty()) {
    policyContext.reportProblem("No groupse assigned...");
    return false;
}

```

Reply:
Regarding the first question: even though I assign an error message to the variable, I just report the error if it is a positive operator, as it is validated on the line 138. 
About the other test scenarios, I will add them here.

Reply:
no, i'm saying that an empty assigned group is _not_ an error, because if the policy states ` eq []`, then a BPN that has no groups assigned to it would satisfy that condition. 

Generally though, even if the code were semantically correct, the control flow is too complicated, there are duplicated fragments.

Reply:
So, in this case, would you agree that both _groups.failed()_ and _groups.getContent().isEmpty()_ validations wouldn't make sense anymore? Since in both cases, we could consider the result as an empty list, then validate directly on the operator evaluators. 

For example, we would have:
- `[] eq []` = true
- `[] isAllOf []` = true
- `[] isAnyOf []` = false
- `[] in []` = false

Reply:
nope, that's incorrect. there is a fundamental difference between a `Result.failure` and an empty list. A `failure` could be caused e.g. by a database error or something. That is an error situation. An empty list is not an error situation, because a particular BPN might simply not have any groups assigned. There could still be constraints that evaluate to `true`. 

I also disagree with your examples, because `[] isAnyOf []` should be true. Any element on the left side (Which is none) is equal to any element of the right side (which is also none). And `[] in []` is also true, because the entire list on the left side (which is empty) is contained fully in the right side (also being empty).

If you have no apples, and I ask you to give me all your apples, and I end up with no apples - did you give me everything you had?

Reply:
Okay, so just this validation `groups.getContent().isEmpty()` would not make sense anymore? 
In regard to the validations from each operator, I will check and fix any inconsistent behavior.

Reply:
@paullatzelsperger I don't know why, but I'm not being able to re-request a review, as you asked me to do after every conversation solved on this PR.